runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -96,20 +147,58 @@ AOTCacheClassRecord::subRecordsDo(const std::function<void(const AOTCacheRecord
    f(_classLoaderRecord);
    }
 
+AOTCacheClassRecord *
+AOTCacheClassRecord::read(FILE *f,
+                          const Vector<AOTCacheClassLoaderRecord *> &classLoaderRecords,
+                          const Vector<AOTCacheClassRecord *> &classRecords,
+                          const Vector<AOTCacheMethodRecord *> &methodRecords,
+                          const Vector<AOTCacheClassChainRecord *> &classChainRecords,
+                          const Vector<AOTCacheWellKnownClassesRecord *> &wellKnownClassesRecords,
+                          const Vector<AOTCacheAOTHeaderRecord *> &aotHeaderRecords)
+   {
+   ClassSerializationRecord header;
+   if (1 != fread(&header, sizeof(header), 1, f))
+      return NULL;
+
+   if (!header.isValid() ||
+       (header.classLoaderId() >= classLoaderRecords.size()))
+      return NULL;
+
+   auto record = new (AOTCacheRecord::allocate(size(header.nameLength()))) AOTCacheClassRecord(classLoaderRecords[header.classLoaderId()]);
```

Comment:
For instance, there is a missing check that `classLoaderRecords[header.classLoaderId()]` is non-null here. The same applies to all other places where we set sub-record pointers.